reference/test/base/reordered.cpp

```diff
@@ -0,0 +1,487 @@
+                                        exec)),
+          // clang-format on
+          diag(Diag::create(exec, 3)),
+          diag2(Diag::create(exec, 5)),
+          diag3(Diag::create(exec, 5)),
+          x(gko::initialize<Vec>({1., 2., 3., 4., 5.}, exec)),
+          b(gko::initialize<Vec>({10.3, 16.5, 11.9, 10., 7.1}, exec)),
+          rcm_factory(Rcm::build().on(exec)),
+          solver_factory(
+              Bicgstab::build()
+                  .with_criteria(
+                      gko::stop::Iteration::build().with_max_iters(100u).on(
+                          exec),
+                      gko::stop::ResidualNorm<value_type>::build()
+                          .with_reduction_factor(r<value_type>::value)
+                          .on(exec))
+                  .on(exec)),
+          tol{r<value_type>::value}
+    {
+        this->create_diag(diag.get());
+        this->create_equilibration(diag2.get(), diag3.get());
+    }
+
+    void create_diag(Diag* d)
+    {
+        auto* v = d->get_values();
+        v[0] = 1.;
+        v[1] = .5;
+        v[2] = .25;
+    }
+
+    void create_equilibration(Diag* d1, Diag* d2)
+    {
+        auto* v1 = d1->get_values();
+        v1[0] = 1.0648;
+        v1[1] = 0.4472;
+        v1[2] = 1.4907;
+        v1[3] = 1.1180;
+        v1[4] = 1.0648;
+        auto* v2 = d2->get_values();
+        v2[0] = 0.4472;
+        v2[1] = 0.4472;
+        v2[2] = 0.4472;
+        v2[3] = 0.4472;
+        v2[4] = 0.4472;
+    }
+
+    std::shared_ptr<const gko::Executor> exec;
+    std::shared_ptr<Mtx> mtx;
+    std::shared_ptr<Mtx> row_scaled;
+    std::shared_ptr<Mtx> col_scaled;
+    std::shared_ptr<Mtx> row_col_scaled;
+    std::shared_ptr<Mtx> rcm_mtx;
+    std::shared_ptr<Mtx> rcm_mtx_reordered;
+    std::shared_ptr<Mtx> rectangular_mtx;
+    std::shared_ptr<Diag> diag;
+    std::shared_ptr<Diag> diag2;
+    std::shared_ptr<Diag> diag3;
+    std::shared_ptr<typename Rcm::Factory> rcm_factory;
+    std::shared_ptr<typename Bicgstab::Factory> solver_factory;
+    std::shared_ptr<Vec> b;
+    std::shared_ptr<Vec> x;
+    gko::remove_complex<value_type> tol;
+};
+
+TYPED_TEST_SUITE(ScaledReordered, gko::test::ValueIndexTypes,
+                 PairTypenameNameGenerator);
+
+
+TYPED_TEST(ScaledReordered, BuildsWithOnlySystemMatrix)
+{
+    using SR = typename TestFixture::SR;
+    using Mtx = typename TestFixture::Mtx;
+    auto scaled_reordered_fact = SR::build().on(this->exec);
+    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
+
+    GKO_ASSERT_MTX_NEAR(gko::as<Mtx>(scaled_reordered->get_system_matrix()),
+                        this->mtx, this->tol);
+}
+
+
+TYPED_TEST(ScaledReordered, BuildsWithRowScaling)
+{
+    using SR = typename TestFixture::SR;
+    using Mtx = typename TestFixture::Mtx;
+    auto scaled_reordered_fact =
+        SR::build().with_row_scaling(this->diag).on(this->exec);
+    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
```

Comment:
nit  (for AAA separation):
```suggestion
        SR::build().with_row_scaling(this->diag).on(this->exec);
        
    auto scaled_reordered = scaled_reordered_fact->generate(this->mtx);
```